src/main/java/com/aws/greengrass/lifecyclemanager/Kernel.java

```diff
@@ -196,24 +199,19 @@ public Kernel launch() {
         switch (deploymentStageAtLaunch) {
             case BOOTSTRAP:
                 logger.atInfo().kv("deploymentStage", deploymentStageAtLaunch).log("Resume deployment");
-                int exitCode;
                 try {
-                    exitCode = bootstrapManager.executeAllBootstrapTasksSequentially(
-                            deploymentDirectoryManager.getBootstrapTaskFilePath());
-                    if (!bootstrapManager.hasNext()) {
-                        logger.atInfo().log("Completed all bootstrap tasks. Continue to activate deployment changes");
-                    }
-                    // If exitCode is 0, which happens when all bootstrap tasks are completed, restart in new launch
-                    // directories and verify handover is complete. As a result, exit code 0 is treated as 100 here.
-                    logger.atInfo().log((exitCode == REQUEST_REBOOT ? "device reboot" : "Nucleus restart")
-                            + " requested to complete bootstrap task");
-
-                    shutdown(30, exitCode == REQUEST_REBOOT ? REQUEST_REBOOT : REQUEST_RESTART);
+                    Path bootstrapTaskFilePath = deploymentDirectoryManager.getBootstrapTaskFilePath();
+                    executeBootstrapTasksAndShutdown(bootstrapManager, bootstrapTaskFilePath);
                 } catch (ServiceUpdateException | IOException e) {
                     logger.atError().log("Deployment bootstrap failed", e);
                     try {
+                        // Bootstrapping for target deployment failed, so check if bootstrap-on-rollback is needed
+                        boolean bootstrapOnRollbackRequired = kernelAlts.prepareBootstrapOnRollbackIfNeeded(
```

Comment:
if a component bootstrap times out in `executeOneBootstrapTask`, it throws a `ServiceUpdateException` directly without persisting bootstrap list. 

I think this case is not handled because you rely on updated bootstrap list to check which component needs roll back.

Reply:
Can we also add a test case for this?

Reply:
I assume you are referring to the possibility that `executeBootstrapTasksAndShutdown` throws a ServiceUpdateException during `executeOneBootstrapTask`, leading into the current context.

Actually, when `executeOneBootstrapTask` fails, the bootstrap task list is never updated!

I am resolving this problem by adding a private member `activeTask` into `BootstrapManager`, which allows me to flag this component for bootstrap-on-rollback. See usage of `BootstrapManager.getPendingTasks()`.

Reply:
I'm going to rename `BootstrapManager.getPendingTasks()` -> `BootstrapManager.getUnstartedTasks()` and add some testing into `BootstrapManagerTest`.